hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/protocolPB/OMInterServiceProtocolServerSideImpl.java

```diff
@@ -60,7 +64,7 @@ public BootstrapOMResponse bootstrap(RpcController controller,
           .build();
     }
 
-    checkLeaderStatus();
+    OzoneManagerRatisUtils.checkLeaderStatus(ozoneManager);
```

Comment:
Why go through this nesting? alternatively we could `ozoneManager.checkLeaderStatus`
```suggestion
    ozoneManager.checkLeaderStatus();
```

Reply:
All protocol service API need throws ServiceException, while ozoneManager.checkLeaderStatus throws OMNotLeaderException and OMLeaderNotReadyException.  So we need OzoneManagerRatisUtils.checkLeaderStatus wrapper to covert exceptions.

ozoneManager.checkLeaderStatus is used both in SASL path authentication and protocol level authentication.  It throws the core exception. Let the caller wrap the core exception into desired outer exceptions.

Reply:
Hi @ChenSammi, thanks for your updated comment on handling the leader exceptions.

> All protocol service API need throws ServiceException, while ozoneManager.checkLeaderStatus throws OMNotLeaderException and OMLeaderNotReadyException. So we need OzoneManagerRatisUtils.checkLeaderStatus wrapper to covert exceptions.

Having that, the `OzoneManagerProtocolServerServerSideTranslatorPB.processRequest` needs to throw a `ServiceException` should `isRatisEnabled == true` and it is _not_ a leader.  There looks to be two cases where this happens in `processRequest`:

1. For the case that `isRatisEnabled == true` and `s3Auth == false`, that's what happens with your leader check `OzoneManagerRatisUtils.checkLeaderStatus`.  It is handled.
2. For the case that `isRatisEnabled == true` and `request.hasS3Authentication == true` it needs to be handled so that the `ServiceException` is thrown in the event of a leadership check exception.  This looks to be currently wrapping the leadership exception in an IOException (`InvalidToken`) and returned to the caller through the error `OmResponse`.  Does that look to be the case?  If so, instead it needs to throw a leader exception wrapped in a `ServiceException` for the caller.  If we are not throwing the `ServiceException`, would you like to insert the `OzoneManagerRatisUtils.checkLeaderStatus` to do so?  Or have the `IOException` handler for `processRequest` (catch block) wrap the extracted` Leader exception` into a `ServiceException` and throw?

Reply:
Hey @neils-dev,  sorry for the late response.  Just want to confirm that the comments left to address is the following, right? 

> The javadoc for validateS3Credential can be updated to include the ServiceException that may be thrown.

Reply:
Thanks @ChenSammi for following up.  Yes, just the javadoc addition for correctness.  LGTM.  Thanks.

Reply:
@neils-dev , the javadoc is updated. Would you like to take another look?

Reply:
Thanks @ChenSammi.  Looks great.  I tested the changes on the secure docker development clusters as well without issues -  clients handle OMException raised as expected for authentication errors (S3) and client failovers for raised ServiceException wrapped OMNotALeader exceptions.

Reply:
Thanks @neils-dev . Do you think the PR can be merged now?

Reply:
LGTM

Reply:
+1